File: src/main/java/com/glitch/knowledge/model/Terminus.java
package com.glitch.knowledge.model;

import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;

/**
 * Map of {@link Line} to any {@link Station} that is a terminus for the line.
 * <br>
 * Lines that are {@link LineType} LOOP should have at least one station
 * randomly chosen as a terminus. This is for compatibility and will not affect
 * the code.
 */
@Entity(name = "terminus")
public class Terminus {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private int id;

    @ManyToOne
    @JoinColumn(name = "line_id", referencedColumnName = "id")
    private Line line;

    @ManyToOne
    @JoinColumn(name = "station_id", referencedColumnName = "id")
    private Station terminus;
}

File: src/main/kotlin/com/glitch/knowledge/model/Terminus.kt
package com.glitch.knowledge.model

import org.jetbrains.exposed.v1.core.dao.id.EntityID
import org.jetbrains.exposed.v1.core.dao.id.IntIdTable
import org.jetbrains.exposed.v1.dao.ImmutableEntityClass
import org.jetbrains.exposed.v1.dao.IntEntity


internal object TerminusTable : IntIdTable("terminus") {
    val lineId = reference("line_id", LineTable.id)
    val stationId = reference("station_id", StationTable.id)
}


/**
 * Map of [Line] to any [Station] that is a terminus for the line.
 *
 * Lines that are [com.glitch.knowledge.model.constant.LineType] LOOP should have at least one station
 * randomly chosen as a terminus. This is for compatibility and will not affect
 * the code.
 *
 * @property id Database id.
 * @property line [Line] that this station terminates.
 * @property terminus The [Station] object for this terminus.
 */
internal class TerminusDAO(id: EntityID<Int>) : IntEntity(id) {
    companion object : ImmutableEntityClass<Int, TerminusDAO>(TerminusTable)

    val line by LineDAO.referencedOn(TerminusTable.lineId)
    val terminus by StationDAO.referencedOn(TerminusTable.stationId)
}